desktop/ui/src/main/java/org/datacleaner/widgets/visualization/JobGraphTransformers.java

```diff
@@ -235,7 +240,20 @@ public Icon transform(Object obj) {
                 } else {
                     configured = componentBuilder.isConfigured(false);
                 }
-                return IconUtils.getDescriptorIcon(descriptor, configured, IconUtils.ICON_SIZE_LARGE);
+
+                Icon descriptorIcon = IconUtils.getDescriptorIcon(descriptor, configured, IconUtils.ICON_SIZE_LARGE);
+
+                if (descriptor instanceof RemoteTransformerDescriptorImpl) {
+                    Map<DescriptorProvider, DescriptorProviderStatus> actualDescriptorProviderStatusMap  = componentBuilder.getAnalysisJobBuilder().getConfiguration().getEnvironment().getDescriptorProvider().getActualStatusMap();
```

Comment:
OK so this answers what the method is used for - but unfortunately also confirms why it's not so good.

This method (```Icon transform(Object)```) is invoked for _every_ change in the graph and for _every_ component on it - even just a move of a node to a new position or whatever. So having to traverse through descriptors, maps and states is maybe a small operation, but you're going to do it again and again and again. This can make our UX become quite slow and annoying.

Which is why a listener approach is much better. When a change is detected (by some background process) - update something. In this case I would suggest to have a listener that is aware of which descriptor providers are in error state. That way you will essentially have something like ```Set<ComponentDescriptor> errorStateDescriptors``` which you can do a single check in: ```if (errorStateDescriptors.contains(descriptor) { addErrorOverlay(icon); }```.